Hardened Eth 1.0 voting strategy (#1218)
ethereum/eth2.0-specs: specs/validator/0_beacon-chain-validator.md

##### Eth1 Data

> The `block.eth1_data` field is for block proposers to vote on recent Eth 1.0 data. This recent data contains an Eth 1.0 block hash as well as the associated deposit root (as calculated by the `get_deposit_root()` method of the deposit contract) and deposit count after execution of the corresponding Eth 1.0 block. If over half of the block proposers in the current Eth 1.0 voting period vote for the same `eth1_data` then `state.eth1_data` updates at the end of the voting period. Each deposit in `block.body.deposits` must verify against `state.eth1_data.eth1_deposit_root`.

Reviewer: I think it is important to note here that `eth1_data` is only updated in times of finality, particularly because not much is said about `eth1_data` in `0_beacon_chain.md`. Speaking with the guys from RV, they initially missed this fact and it lead to some concerns about safety.

Reviewer 2: It is not only updated in times of finality. 

I was mistaken yesterday that incoming deposits are only activate at times of finality, but even this is not true. I'd leave as is for now, but will think about the scenario discussed yesterday a bit more.

Reviewer 2: Trying to grok [`process_registry_updates`](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#registry-updates) about the activations at finailty.

regardless. eth1data is updated whenever possible. The thing that might protect against the scenario discussed yesterday will have to do with activations.

Author: > it lead to some concerns about safety

What concerns of safety?

Reviewer 2: Can discuss offline. Nothing that should be addressed in this PR

Let `get_eth1_data(distance: int) -> Eth1Data` be the (subjective) function that returns the Eth 1.0 data at distance `distance` relative to the Eth 1.0 head at the start of the current Eth 1.0 voting period. Let `previous_eth1_distance` be the distance relative to the Eth 1.0 block corresponding to `state.eth1_data.block_hash` at the start of the current Eth 1.0 voting period. An honest block proposer sets `block.eth1_data = get_eth1_vote(state, previous_eth1_distance)` where:

```python
def get_eth1_vote(state: BeaconState, previous_eth1_distance: uint64) -> Eth1Data:
    # Get fresh and stale Eth 1.0 data

>     fresh_eth1_data = list(map(get_eth1_data, range(ETH1_FOLLOW_DISTANCE, 2 * ETH1_FOLLOW_DISTANCE)))

Reviewer 2: We have tended to use list comprehensions in favor of maps throughout the spec.
```suggestion
    fresh_eth1_data =[get_eth1_data(distance) for distance in range(ETH1_FOLLOW_DISTANCE, 2 * ETH1_FOLLOW_DISTANCE)]
```

>     stale_eth1_data = list(map(get_eth1_data, range(2 * ETH1_FOLLOW_DISTANCE, previous_eth1_distance)))

Reviewer 2: ```suggestion
    stale_eth1_data = [get_eth1_data(distance) for distance in range(2 * ETH1_FOLLOW_DISTANCE, previous_eth1_distance)]
```

    # Filter valid votes from the current voting period
    valid_votes = []

>     for slot, vote in enumerate(state.eth1_data_votes):

Reviewer 2: suggest a few minor changes for readability and line length reduction

```python
    fresh_eth1_data = set(get_eth1_data(distance) for distance in range(ETH1_FOLLOW_DISTANCE, 2 * ETH1_FOLLOW_DISTANCE))
    state_eth1_data = set(get_eth1_data(distance) for distance in range(2 * ETH1_FOLLOW_DISTANCE, previous_eth1_distance))
    all_eth1_data = fresh_eth1_data.union(state_eth1_data)

    valid_votes = []
    for slot, vote in enumerate(state.eth1_data_votes):
        start_of_period = slot % SLOTS_PER_ETH1_VOTING_PERIOD < integer_square_root(SLOTS_PER_ETH1_VOTING_PERIOD
        if start_of_period and vote in fresh_eth1_data:
            valid_votes.append(vote)
        elif not start_of_period and vote in all_eth1_data:
            valid_votes.append(vote)
    return max(valid_votes, key=valid_votes.count, default=get_eth1_data(ETH1_FOLLOW_DISTANCE))
```

Author: Great suggestion. Made it *even* shorter and readable :)

        if slot % SLOTS_PER_ETH1_VOTING_PERIOD < integer_square_root(SLOTS_PER_ETH1_VOTING_PERIOD):
            if vote in set(fresh_eth1_data):
                valid_votes.append(vote)
        else:
            if vote in set(fresh_eth1_data).union(stale_eth1_data):
                valid_votes.append(vote)

>     return max(valid_votes, valid_votes.count, default=get_eth1_data(ETH1_FOLLOW_DISTANCE))

Reviewer: As far as I know, the key over which the max operates should be passed in as a keyword argument. (And if this is not the case, I think my proposed change makes it more readable regardless.)

```suggestion
    return max(valid_votes, key=valid_votes.count, default=get_eth1_data(ETH1_FOLLOW_DISTANCE))
```
